### PR TITLE
feat: display survey on generic actions

### LINKED ISSUE #391: [FEATURE] Display survey on generic actions
## Name: Display on generic actions

**Is your feature request related to a problem? Please describe.**
As a user I have to set up specific user actions to be able to trigger a survey. This is quite a hurdle which lowers user activation.

**Describe the solution you'd like**
As a user I want to trigger surveys on generic user actions:

- Scroll depth and
- Exit intent (on desktop)

**Tasks**

Part 1:

These triggers are listened to in the JS Widget at packages/js. Add two new listeners to the widget:

1. Exit intent
2. Scroll depth. You can set a default of 50%, down the line we will allow users to set a custom scroll depth in %.

Part 2: 

When a Product gets created, the Actions and Attribtues are created with it. In the same way we have Automatic Attributes (email and userID) we will create Automatic Actions on Product Creation. Currently, a Product is created on three occasions:

1. Add new team → adds new product
2. Google + GitHub Auth → adds new product
3. Create user (Email) → adds new product

Add two Automatic Actions in the Product Creation:

1. **Exit Intent**
Description: A user on Desktop leaves the website with the cursor.
2. 50% Scroll
Description: A user scrolled 50% of the current page

Part 3: Test it 😊

## How to code for Formbricks
1. Make sure everything is type-safe
2. If you run `pnpm dev` you'll find a demo app to test in-app surveys at `localhost:3002`
3. When your code is done and working, paste it to chatGPT and ask for refactoring ideas. Apply what makes sense to you (GPT-4 is a lot better). Try out our new [Formbricks-Prompt](https://github.com/formbricks/formbricks/blob/main/megaprompt.md) ✨

### PR DESCRIPTION
**Pull request by Ashutosh-Bhadauriya**
## What does this PR do?

This PR adds two more automatic actions, Exit Intent and Scroll Depth 50%

Fixes #391

Screenshot

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Go to actions. here you can see these two automatic actions
- Open the demo app, you can check exit intent here and in order to check scroll depth 50%, we need to add more content in demo, which can be dummy lorem ipsum text

## Checklist

- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
@Ashutosh-Bhadauriya Great work, works smoothly! 🙏 

@mattinannt Please do a final review

**Review by mattinannt — CHANGES_REQUESTED** [review]
Feature works great, thanks 😊🎉

When an existing user (an existing product) uses the formbricks-js widget and the exit-event or 50%-event is triggered, a new eventClass is created by the client's event endpoint ( `apps/web/pages/api/v1/client/environments/[environmentId]/events/index.ts` ). This public endpoint is only allowed to create `code` events, so the resulting event won't be an automatic event, but a code event. As a workaround, we can add a name check to this endpoint that will create an automatic event for an eventClass named `Exit Intent (Desktop)` or `50% Scroll` and a code event otherwise. It's a bit hacky, but I think it's the easiest way to implement a (temporary) data migration.
Do you want to make these changes to the code as well as a small clean-up in the formbricks-js init function as suggested below? 🤗

After that we are ready to merge the changes into the main 😊💪.

File: packages/js/src/lib/init.ts [hunk 1]
```diff
@@ -127,11 +127,47 @@ export const initialize = async (
   logger.debug("Add click event listeners");
   addClickEventListener();
 
-  logger.debug("Initialized");
+  logger.debug("Add exit intent (Desktop) listener");
+  if(typeof document !== undefined){
```
[the remark this answers is not in the document]
Comment (Ashutosh-Bhadauriya) [reply packages/js/src/lib/init.ts:131]: @mattinannt sure, let me fix these 🙂

**Comment by Ashutosh-Bhadauriya** [comment]
@mattinannt just pushed the changes. Please let me know😊

**Review by mattinannt — APPROVED** [review]
@Ashutosh-Bhadauriya Sorry that it took a bit longer to review the changes.
Thanks a lot for the PR and the updates, everything works great and package-js code looks good & clean now with the new automaticEvents functions 😊🎉💪

I just updated the formatting and merged the latest changes from main to make sure everything works fine. We use prettier to format the code and the easiest way to comply with the formatting rules is to use the [VSCode plugin](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) and enable formatOnSave 😊

I will merge the changes into main now 💪

**Comment by Ashutosh-Bhadauriya** [comment]
Thanks @mattinannt 😊 Actually I use prettier for formatting, but it stopped working after I moved to WSL. Sorry for the issue. I have fixed it now🙂

**Pull request merged**